lib/concurrent/configuration.rb

```diff
@@ -1,16 +1,21 @@
 require 'thread'
-require 'concurrent/atomics'
 require 'concurrent/errors'
-require 'concurrent/executors'
 require 'concurrent/concern/deprecation'
 require 'concurrent/concern/logging'
-require 'concurrent/utility/at_exit'
-require 'concurrent/utility/processor_counter'
 
 module Concurrent
   extend Concern::Logging
   extend Concern::Deprecation
 
+  autoload :AtExit,             'concurrent/utility/at_exit'
+  autoload :AtomicReference,    'concurrent/atomic/atomic_reference'
+  autoload :Delay,              'concurrent/delay'
+  autoload :ProcessorCounter,   'concurrent/utility/processor_counter'
+  autoload :TimerSet,           'concurrent/executor/timer_set'
+  autoload :ImmediateExecutor,  'concurrent/executor/immediate_executor'
```

Comment:
We could also clear this up by creating a boot (or core) file which will always load the minimal required files for the concurrent-ruby to function it would load these problematic files. All the other abstractions would load boot first and then any other required abstractions. This way the circular dep. would be always handled by boot.